### PR TITLE
feat(wire): ensure prop type matches adapter value @W-16314799@

### PR DESCRIPTION
**Pull request by wjhsf**
## Details

When using a wire adapter (e.g. `@wire(getData) data: Type`), the type of the decorator returned by `@wire()` does not include any type information from the adapter provided, and therefore no type checking can be done to ensure that the adapter is providing the type that the prop expects. For example, `@wire(getString) prop: number` would not result in a type error, even though it should. This PR fixes that.

This also provides limited type checking that the config passed to `@wire` matches the config expected by the adapter. If the value of a prop is a string (e.g. `@wire(getData, {val: '$prop'})`), we know that the value used by the adapter is a value from the component, so we opt out of type checking strings.<sup>1</sup> But all other values work as expected.

### Known limitations

* Due to the way TypeScript works, the type of the prop cannot be inferred from the decorator (it's the other way around - the decorator type is derived from the prop).
* TypeScript doesn't know about the LWC component lifecycle. It expects all properties to be initialized in the constructor or with a default value, unless `undefined` is a valid type. In practice, this means that all `@wire`-decorated properties will need to use a non-null assertion (e.g. `@wire(getData) data!: Type`), mark the property as optional, or assign a default value that gets overwritten.

<sup>1</sup> If we decide to drop support for experimental decorators, modern decorators have access to the component type, so we _could_ resolve reactive props to the correct type from the component.

## Does this pull request introduce a breaking change?

-   💔 Yes, it does introduce a breaking change.

If you're using `@wire` incorrectly in a TypeScript project, you'll now see a type error.

## Does this pull request introduce an observable change?

-   🔬 Yes, it does include an observable change.

## GUS work item

W-16314799

### CODE REVIEW

File: .husky/pre-push [hunk 1]
```diff
@@ -1,4 +1,3 @@
-#!/usr/bin/env sh
```
Comment (nolanlawson) [inline .husky/pre-push:1]: merge with `master` and this will go away

**Review by nolanlawson — COMMENTED** [review]
LGTM but let's test on core to see what kinds of breakages to expect. I'll do a canary release.

**Comment by nolanlawson** [comment]
lwc `7.3.0-alpha.0` has this

**Comment by nolanlawson** [comment]
Also downstream Nucleus repos are failing – anything we can do proactively to fix them?

**Comment by wjhsf** [comment]
> Also downstream Nucleus repos are failing – anything we can do proactively to fix them?

I was gonna cut a prerelease today to investigate the failures. Thanks for doing that first part!

**Comment by wjhsf** [comment]
Replaced by #4429

**Pull request closed without merging**